docs/proposals/20200602-machine-deletion-phase-hooks.md

```diff
@@ -0,0 +1,289 @@
+The machine-controller should check for the existence of 1 or more hooks at
+specific points (lifecycle-points) during reconciliation.  If a hook matching
+the lifecycle-point is discovered, the machine-controller should stop
+reconciling the machine.
+
+An example of where the pre-drain lifecycle-point might be implemented:
+https://github.com/kubernetes-sigs/cluster-api/blob/30c377c0964efc789ab2f3f7361eb323003a7759/controllers/machine_controller.go#L270
+
+##### Reconciliation
+When an Owning Controller updates the machine, reconciliation will be
+triggered, and the machine will continue reconciling as normal, unless another
+hook is still present; there is no need to 'fail' the reconciliation to
+enforce requeuing.
+
+When all hooks for a given lifecycle-point are removed, reconciliation
+will continue as normal.
+
+##### Hook failure
+The machine-controller should not timeout or otherwise consider the lifecycle
+hook as 'failed.'  Only the Owning Controller may decide to remove a
+particular lifecycle hook to allow the machine-controller to progress past the
+corresponding lifecycle-point.
+
+##### Hook ordering
+The machine-controller will not attempt to enforce any ordering of hooks.
+
+Owning Controllers may devise a dependency-based enforcement mechanism
+via whatever means Owning Controllers determine.  Examples could be
+using CRDs external to the machine-api, gRPC communications, or
+additional annotations on the machine or other objects.
+
+#### Owning Controller Design
+Owning Controller is the component that manages a particular lifecycle hook.
+
+##### Owning Controllers must
+* Watch machine objects and determine when an appropriate action must be taken.
+* After completing the desired hook action, remove the hook annotation.
+##### Owning Controllers may
+* Watch machine objects and add a hook annotation as desired by the operator
+* After completing a hook, set an addition annotation that indicates this
+controller has finished for dependency ordering, or perform some other action
+to signal other Owning Controllers that they may proceed.
+
+#### Determining when to take action
+
+An Owning Controller should watch machines and determine when is the
+best time to take action.
+
+For example, if an Owning Controller manages a lifecycle hook at the
+pre-drain lifecycle-point, then that controller should take action
+immediately after a machine has a DeletionTimestamp or enters the
+"Deleting" phase.
+
+Fine-tuned coordination is not possible at this time; eg, it's not
+possible to execute a pre-term hook only after a node has been drained.
+
+##### Failure Mode
+It is entirely up to the owning controller to determine when it is prudent to
+remove a particular lifecycle hook.  Some controllers may want to 'give up'
+after a certain time period, and others may want to block indefinitely.
+Cluster operators should consider the characteristics of each controller before
+utilizing them in their clusters.
```

Comment:
it seems likely that an owning controller could remove its hook annotation if it reaches some "give up" failure state. it does make me wonder if we need any sort of mechanism to bubble those errors back through the capi controller?

i'm guessing not, it just seems like too much coordination between disparate components but i wanted to ask.

Reply:
> it seems likely that an owning controller could remove its hook annotation if it reaches some "give up" failure state.

That's what I was implying, I can be more clear.

> it does make me wonder if we need any sort of mechanism to bubble those errors back through the capi controller?

For at least deletion hooks that give up, that state is likely to be short lived on the machine-object itself.

It's not clear what the abstraction should be for reporting failures.  I think the machine being stuck is reporting enough.  This seems like it would be a proposal unto itself.

Reply:
I'd say for now we either do nothing, or we have the machine controller add a new condition that indicates it's waiting on hooks. I don't think we need to block proceeding on this, though.